File: funcional/src/Agenda.hs
module Agenda where

-- Imports
import System.IO
import System.Directory
import qualified Data.Text as T
import Data.Typeable

-- Construtor
data Agenda = Agenda {
    dia :: String,
    visitas_8_as_10 :: Int,
    visitas_10_as_12 :: Int,
    visitas_12_as_14 :: Int,
    visitas_14_as_16 :: Int,
    visitas_16_as_18 :: Int
} deriving (Show, Read)

---------------------------Criar uma Agenda-------------------------------------

criarAgenda :: String -> IO()
criarAgenda dia = do
    appendFile "../db/agendas.txt" (dia ++ ", " ++ (show 0) ++ ", " ++ (show 0) ++ ", " ++ (show 0) ++ ", " ++ (show 0) ++ ", " ++ (show 0) ++ "\n")
    return ()

cadastrarAgenda :: String -> String -> String -> String -> String -> String -> IO()
cadastrarAgenda dia visitas1 visitas2 visitas3 visitas4 visitas5 = do
    appendFile "../db/agendas.txt" (dia ++ ", " ++ visitas1 ++ ", " ++ visitas2 ++ ", " ++ visitas3 ++ ", " ++ visitas4 ++ ", " ++ visitas5 ++ "\n")
    return ()

---------------------------Buscar uma Agenda------------------------------------

buscaUmaAgenda:: String -> IO()
buscaUmaAgenda diaDado = do
    agendas <- readFile "../db/agendas.txt"
    let listaAgendas = T.splitOn (T.pack "\n") (T.pack agendas)
    let agendasConvertidas = [ toAgenda (T.unpack agenda) | agenda <- listaAgendas]
    let listaEncontrada = [ agenda | agenda <- agendasConvertidas, dia agenda == diaDado]
    if (length listaEncontrada == 0) then print ("Agenda nao encontrada")
    else print (listaEncontrada !! 0)

---------------------------Listar Agendas---------------------------------------

listarAgendas :: IO()
listarAgendas = do
    handle <- openFile "../db/agendas.txt" ReadMode  
    conteudo <- hGetContents handle
    let listaQuebrada = T.splitOn (T.pack "\n") (T.pack conteudo)
    let listaConvertida = [ toAgenda (T.unpack agenda) | agenda <- listaQuebrada, agenda /= (T.pack "")]
    putStr (listaToString listaConvertida)
    hClose handle

listaToString :: [Agenda] -> String
listaToString [] = []
listaToString (e:es) = if not (null es) then do toString e ++ "\n\n" ++ listaToString es
    else do toString e ++ "\n"

---------------------------Remover uma Agenda-----------------------------------

removerAgenda :: String -> IO()
removerAgenda diaDado = do
    handle <- openFile "../db/agendas.txt" ReadMode  
    tempdir <- getTemporaryDirectory  
    (tempName, tempHandle) <- openTempFile tempdir "temp"  
    contents <- hGetContents handle
    let listaComAgendas = lines contents
    let agendasResultantes = [agenda | agenda <- listaComAgendas, (dia (toAgenda agenda)) /= diaDado]
    let agendasFormatadas = map (T.unpack . T.toTitle . T.pack) agendasResultantes
    hPutStr tempHandle $ unlines agendasFormatadas 
    hClose handle  
    hClose tempHandle  
    removeFile "../db/agendas.txt" 
    renameFile tempName "../db/agendas.txt"

---------------------------Editar uma Agenda------------------------------------

editarAgenda :: String -> String -> String -> String -> String -> String -> String -> IO()
editarAgenda dia novoDia visitas1 visitas2 visitas3 visitas4 visitas5 = do
    removerAgenda dia
    cadastrarAgenda novoDia visitas1 visitas2 visitas3 visitas4 visitas5
    return ()

---------------------------Agendar uma visita-----------------------------------

agendarVisita :: String -> IO()
agendarVisita diaEscolhido = do
    agendas <- readFile "../db/agendas.txt"
    let listaAgendas = T.splitOn (T.pack "\n") (T.pack agendas)
    let agendasConvertidas = [ toAgenda (T.unpack agenda) | agenda <- listaAgendas]
    let listaEncontrada = [ agenda | agenda <- agendasConvertidas, dia agenda == diaEscolhido]
    if ((length listaEncontrada) == 0) then do
        cadastrarAgenda diaEscolhido "1" "0" "0" "0" "0"
        print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 8h as 10h")
    else do
        let agenda = (listaEncontrada !! 0)
        if ((visitas_8_as_10 agenda) < 30) then do
            editarAgenda diaEscolhido diaEscolhido (show ((visitas_8_as_10 agenda) + 1)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 8h as 10h")
        else if((visitas_10_as_12 agenda) < 30) then do
            editarAgenda diaEscolhido diaEscolhido (show (visitas_8_as_10 agenda)) (show ((visitas_10_as_12 agenda) + 1)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 10h as 12h")
        else if((visitas_12_as_14 agenda) < 30) then do
            editarAgenda diaEscolhido diaEscolhido (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show ((visitas_12_as_14 agenda) + 1)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 12h as 14h")
        else if((visitas_14_as_16 agenda) < 30) then do
            editarAgenda diaEscolhido diaEscolhido (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show ((visitas_14_as_16 agenda) + 1)) (show (visitas_16_as_18 agenda))
            print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 14h as 16h")
        else if((visitas_16_as_18 agenda) < 30) then do
            editarAgenda diaEscolhido diaEscolhido (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show ((visitas_16_as_18 agenda) + 1))
            print ("\nVoce foi agendado para " ++ diaEscolhido ++ " entre 16h as 18h")
        else print ("Nao temos mais vagas para este dia!")

---------------------------Cancelar uma Visita----------------------------------

cancelarUmaVisita :: String -> String -> IO()
cancelarUmaVisita diaAgendado horario = do
    agendas <- readFile "../db/agendas.txt"
    let listaAgendas = T.splitOn (T.pack "\n") (T.pack agendas)
    let agendasConvertidas = [ toAgenda (T.unpack agenda) | agenda <- listaAgendas]
    let listaEncontrada = [ agenda | agenda <- agendasConvertidas, dia agenda == diaAgendado]
    if ((length listaEncontrada) == 0) then do
        print "\nNao tem nenhum agendamento para este dia."
    else do
        let agenda = (listaEncontrada !! 0)
        if ((visitas_8_as_10 agenda) > 0 && horario == "8h as 10h") then do
            editarAgenda diaAgendado diaAgendado (show ((visitas_8_as_10 agenda) - 1)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nAgendamento para " ++ diaAgendado ++ " entre 8h as 10h foi cancelado")
        else if((visitas_10_as_12 agenda) > 0 && horario == "10h as 12h") then do
            editarAgenda diaAgendado diaAgendado (show (visitas_8_as_10 agenda)) (show ((visitas_10_as_12 agenda) - 1)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nAgendamento para " ++ diaAgendado ++ " entre 10h as 12h foi cancelado")
        else if((visitas_12_as_14 agenda) > 0 && horario == "12h as 14h") then do
            editarAgenda diaAgendado diaAgendado (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show ((visitas_12_as_14 agenda) - 1)) (show (visitas_14_as_16 agenda)) (show (visitas_16_as_18 agenda))
            print ("\nAgendamento para " ++ diaAgendado ++ " entre 12h as 14h foi cancelado")
        else if((visitas_14_as_16 agenda) > 0 && horario == "14h as 16h") then do
            editarAgenda diaAgendado diaAgendado (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show ((visitas_14_as_16 agenda) - 1)) (show (visitas_16_as_18 agenda))
            print ("\nAgendamento para " ++ diaAgendado ++ " entre 14h as 16h foi cancelado")
        else if((visitas_16_as_18 agenda) > 0 && horario == "16h as 18h") then do
            editarAgenda diaAgendado diaAgendado (show (visitas_8_as_10 agenda)) (show (visitas_10_as_12 agenda)) (show (visitas_12_as_14 agenda)) (show (visitas_14_as_16 agenda)) (show ((visitas_16_as_18 agenda) - 1))
            print ("\nAgendamento para " ++ diaAgendado ++ " entre 16h as 18h foi cancelado")
        else print ("\nNao tem nenhum agendamento para este horario.")

---------------------------Menus------------------------------------------------

menuCadastrarAgenda :: IO()
menuCadastrarAgenda = do
    putStr "\nQual data você gostaria de cadastrar uma agenda? [dd/MM/yyyy]: "
    dia <- getLine
    if (verificaSeDataOk dia) then do
        existe <- verificaSeAgendaExiste dia
        if (existe) then print ("Essa agenda ja existe no banco.")
        else do
            criarAgenda dia
            print "Agenda criada!"
    else print "Esse nao e um formato de data"


menuRemoverAgenda :: IO()
menuRemoverAgenda = do
    putStr "\nQual a data da agenda que você pretende remover? [dd/MM/yyyy]: "
    dia <- getLine
    if (verificaSeDataOk dia) then do
        existe <- verificaSeAgendaExiste dia
        if (existe) then do
            removerAgenda dia
            print "Agenda removida!"
        else print ("Essa agenda nao existe no banco.")
    else print "Esse nao e um formato de data"

menuEditarAgenda :: IO()
menuEditarAgenda = do
    putStr "Informe a data da Agenda que vai ser editada: [dd/MM/yyyy]: "
    dia <- getLine
    if (verificaSeDataOk dia) then do
        existe <- verificaSeAgendaExiste dia
        if (existe) then do
            putStr "Informe a nova data da Agenda que vai ser editada: [dd/MM/yyyy]: "
            novoDia <- getLine
            if (verificaSeDataOk dia) then do
                novoExiste <- verificaSeAgendaExiste novoDia
                if (novoExiste && novoDia /= dia) then print "Essa data ja existe no banco"
                else do
                    putStr "Informe a quantidade de visitas das 08h as 10h: [0 a 30]: "
                    visitas1 <- getLine
                    putStr "Informe a quantidade de visitas das 10h as 12h: [0 a 30]: "
                    visitas2 <- getLine
                    putStr "Informe a quantidade de visitas das 12h as 14h: [0 a 30]: "
                    visitas3 <- getLine
                    putStr "Informe a quantidade de visitas das 14h as 16h: [0 a 30]: "
                    visitas4 <- getLine
                    putStr "Informe a quantidade de visitas das 16h as 18h: [0 a 30]: "
                    visitas5 <- getLine
                    editarAgenda dia novoDia visitas1 visitas2 visitas3 visitas4 visitas5
                    print "Agenda editada!"
            else print "Essa data nao e valida" 
        else print "Essa agenda nao existe no bando para ser editada"
    else print "Essa data nao e valida"

menuAgendarVisita :: IO()
menuAgendarVisita = do
    putStr "Que data você pretende visitar nossa clínica? [dd/MM/yyyy]: "
    dia <- getLine
    if (verificaSeDataOk dia) then do agendarVisita dia
    else print "Essa data nao e valida"

---------------------------Métodos Úteis----------------------------------------

toString :: Agenda -> String
toString (Agenda {dia = d, visitas_8_as_10 = visitas1, visitas_10_as_12 = visitas2, visitas_12_as_14 = visitas3, visitas_14_as_16 = visitas4, visitas_16_as_18 = visitas5}) = d ++ "\n | De 8h às 10h: " ++ (show visitas1) ++ " visitas\n | De 10h às 12: " ++ (show visitas2) ++ " visitas\n | De 12h às 14: " ++ (show visitas3) ++ " visitas\n | De 14h às 16: " ++ (show visitas4) ++ " visitas\n | De 16h às 18: " ++ (show visitas5) ++ " visitas"

toAgenda :: String -> Agenda
toAgenda agendaString = do
    let agendaInList = T.splitOn (T.pack ", ") (T.pack agendaString)
    Agenda {dia = (T.unpack (agendaInList !! 0)), visitas_8_as_10 = (read (T.unpack (agendaInList !! 1))::Int), visitas_10_as_12 = (read (T.unpack (agendaInList !! 2))::Int), visitas_12_as_14 = (read (T.unpack (agendaInList !! 3))::Int), visitas_14_as_16 = (read (T.unpack (agendaInList !! 4))::Int), visitas_16_as_18 = (read (T.unpack (agendaInList !! 5))::Int) }

verificaSeDataOk :: String -> Bool
verificaSeDataOk dataDada = 
    if (length dataList == 3) then do
        if(length (T.unpack (dataList !! 0)) == 2) then do
            if(length (T.unpack (dataList !! 1)) == 2) then do
                if(length (T.unpack (dataList !! 2)) == 4) then do True
                else False
            else False
        else False
    else False
    where
        dataList = T.splitOn (T.pack "/") (T.pack dataDada)

verificaSeAgendaExiste :: String -> IO(Bool)
verificaSeAgendaExiste diaDado = do
    agendas <- readFile "../db/agendas.txt"
    let listaAgendas = T.splitOn (T.pack "\n") (T.pack agendas)
    let agendasConvertidas = [ toAgenda (T.unpack agenda) | agenda <- listaAgendas]
    let listaEncontrada = [ agenda | agenda <- agendasConvertidas, dia agenda == diaDado]
    if ((length listaEncontrada) == 0) then do
        return (False)
    else do
        return (True)

File: logico/src/Agenda.pl
:-use_module(library(csv)).
:-include('Util.pl').

%---------------------------- Cadastrar uma Agenda -----------------------------------

cadastrarAgenda(Data, Horario1, Horario2, Horario3, Horario4, Horario5) :-
    open('../db/Agendas.csv', append, File),
    writeln(File, (Data, Horario1, Horario2, Horario3, Horario4, Horario5)),                 
    close(File).

%---------------------------- Buscar uma Agenda ----------------------------

buscarAgenda(Data, Encontrada) :-
    (exists_file('../db/Agendas.csv') -> (lerCsvRowList('../db/Agendas.csv', Agendas),
    buscaAgendaEmLista(Data, Agendas, Encontrada));
    Encontrada = []).

buscaAgendaEmLista(_, [], []).
buscaAgendaEmLista(Data, [Atual | Restante], Encontrada) :-
    nth0(0, Atual, X),
    (atom_string(X, Data) -> Encontrada = Atual;
    buscaAgendaEmLista(Data, Restante, Encontrada)).
    
%---------------------------- Listar Agendas ---------------------------

listarAgendas :-
    lerCsvRowList('../db/Agendas.csv', Agendas),
    printaAgenda(Agendas).

printaAgenda([]).
printaAgenda([Agenda | Restante]) :- 
    toString(Agenda, String),
    writeln(String),
    printaAgenda(Restante).

%---------------------------- Remover uma Agenda ---------------------------

removerAgenda(Data) :-
    lerCsvRowList('../db/Agendas.csv', Agendas),
    buscarAgenda(Data, Encontrada),
    delete(Agendas, Encontrada, AgendasResultantes),
    delete_file('../db/Agendas.csv'),
    open('../db/Agendas.csv', append, File),
    insereAgendas(AgendasResultantes),
    close(File).

insereAgendas([]).
insereAgendas([Agenda | Restante]) :-
    nth0(0, Agenda, Data),
    nth0(1, Agenda, Horario1),
    nth0(2, Agenda, Horario2),
    nth0(3, Agenda, Horario3),
    nth0(4, Agenda, Horario4),
    nth0(5, Agenda, Horario5),
    cadastrarAgenda(Data, Horario1, Horario2, Horario3, Horario4, Horario5),
    insereAgendas(Restante).

%---------------------------- Editar uma Agenda ---------------------------

editarAgenda(Data, NovaData, Horario1, Horario2, Horario3, Horario4, Horario5) :-
    removerAgenda(Data),
    cadastrarAgenda(NovaData, Horario1, Horario2, Horario3, Horario4, Horario5).

%---------------------------- Agendar Visita ---------------------------

agendarVisita(Data, Horario) :-
    buscarAgenda(Data, Encontrada),
    proper_length(Encontrada, Tamanho),
    (Tamanho =:= 0 -> (cadastrarAgenda(Data, 1, 0, 0, 0, 0),
                       Horario = '08h as 10h');
        (nth0(1, Encontrada, Horario1),
        nth0(2, Encontrada, Horario2),
        nth0(3, Encontrada, Horario3),
        nth0(4, Encontrada, Horario4),
        nth0(5, Encontrada, Horario5),
        (Horario1 < 30 -> (Reserva is Horario1 + 1,
                          editarAgenda(Data, Data, Reserva, Horario2, Horario3, Horario4, Horario5),
                          Horario = '08h as 10h');
        Horario2 < 30 -> (Reserva is Horario2 + 1,
                          editarAgenda(Data, Data, Horario1, Reserva, Horario3, Horario4, Horario5),
                          Horario = '10h as 12h');
        Horario3 < 30 -> (Reserva is Horario3 + 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Reserva, Horario4, Horario5),
                          Horario = '12h as 14h');
        Horario4 < 30 -> (Reserva is Horario4 + 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Horario3, Reserva, Horario5),
                          Horario = '14h as 16h');
        Horario5 < 30 -> (Reserva is Horario5 + 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Horario3, Horario4, Reserva),
                          Horario = '16h as 18h');
        Horario = '0'))).

%---------------------------- Cancelar uma Visita ---------------------------

cancelarVisita(Data, Horario, Resultado) :-
    buscarAgenda(Data, Encontrada),
    proper_length(Encontrada, Tamanho),
    (Tamanho =:= 0 -> writeln('\nNao nenhuma agenda para este dia!');
        (nth0(1, Encontrada, Horario1),
        nth0(2, Encontrada, Horario2),
        nth0(3, Encontrada, Horario3),
        nth0(4, Encontrada, Horario4),
        nth0(5, Encontrada, Horario5),
        ((Horario = "08h as 10h", Horario1 > 0) -> (Reserva is Horario1 - 1,
                          editarAgenda(Data, Data, Reserva, Horario2, Horario3, Horario4, Horario5),
                          Resultado = true);
        (Horario = "10h as 12h", Horario2 > 0) -> (Reserva is Horario2 - 1,
                          editarAgenda(Data, Data, Horario1, Reserva, Horario3, Horario4, Horario5),
                          Resultado = true);
        (Horario = "12h as 14h", Horario3 > 0) -> (Reserva is Horario3 - 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Reserva, Horario4, Horario5),
                          Resultado = true);
        (Horario = "14h as 16h", Horario4 > 0) -> (Reserva is Horario4 - 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Horario3, Reserva, Horario5),
                          Resultado = true);
        (Horario = "16h as 18h", Horario5 > 0) -> (Reserva is Horario5 - 1,
                          editarAgenda(Data, Data, Horario1, Horario2, Horario3, Horario4, Reserva),
                          Resultado = true);
        Resultado = false))).

%---------------------------- Menus ---------------------------

menuBuscarAgenda :-
    (seNaoTemAgendas -> write('Nao ha Agendas cadastradas para buscar!');
    (write('\n'),
    getString(Data, 'Qual a Data da Agenda que vamos buscar? [dd/MM/yyyy]: '),
    buscarAgenda(Data, Encontrada),
    writeln(Encontrada))).

menuListarAgendas :-
    (seNaoTemAgendas -> write('Nao ha Agendas cadastradas para listar!');
    (writeln('\nEsses sao as Agendas no sistema: \n'),
    listarAgendas)).

menuEditarAgenda :-
    (seNaoTemAgendas -> writeln('Nao ha Agendas cadastradas para editar!');
    (write('\n'),
    getString(Data, 'Qual a Data da Agenda que vai ser editada? [dd/MM/yyyy]: '),
    buscarAgenda(Data, Encontrada),
    proper_length(Encontrada, Tamanho),
    (Tamanho =:= 0 -> writeln('Essa Agenda nao existe no sistema!');
        (getString(NovaData, 'Qual a Nova Data para essa Agenda? [dd/MM/yyyy]: '),
        buscarAgenda(NovaData, Encontrada2),
        proper_length(Encontrada2, Tamanho2),
        ((Tamanho2 > 0, Data \= NovaData) -> writeln('Ja existe uma agenda com essa Data!');
            (getString(Horario1, 'Informe a quantidade de visitas das 08h as 10h: [0 a 30]: '),
            getString(Horario2, 'Informe a quantidade de visitas das 10h as 12h: [0 a 30]: '),
            getString(Horario3, 'Informe a quantidade de visitas das 12h as 14h: [0 a 30]: '),
            getString(Horario4, 'Informe a quantidade de visitas das 14h as 16h: [0 a 30]: '),
            getString(Horario5, 'Informe a quantidade de visitas das 16h as 18h: [0 a 30]: '),
            editarAgenda(Data, NovaData, Horario1, Horario2, Horario3, Horario4, Horario5),
            writeln('\nAgenda Editada!'))))))).

menuRemoverAgenda :-
    (seNaoTemAgendas -> write('Nao ha Agendas cadastradas para remover!');
    (write('\n'),
    getString(Data, 'Qual a Data da Agenda que vai ser removida? [dd/MM/yyyy]: '),
    buscarAgenda(Data, Encontrada),
    proper_length(Encontrada, Tamanho),
    (Tamanho =:= 0 -> writeln('Agenda nao existe no sistema!');
        removerAgenda(Data),
        write('Agenda Removida!')))).

menuCadastrarAgenda :-
    write('\n'),
    getString(Data, 'Qual a Data que vai referenciar a Agenda? [dd/MM/yyyy]: '),
    buscarAgenda(Data, Encontrada),
    proper_length(Encontrada, Tamanho),
    (Tamanho > 0 -> writeln('Ja existe uma Agenda com essa data!'); 
        (cadastrarAgenda(Data, 0, 0, 0, 0, 0),
        writeln('Agenda Cadastrada!'))).

menuAgendarVisita :-
    write('\n'),
    getString(Data, 'Que data voce pretende visitar nossa clinica? [dd/MM/yyyy]: '),
    agendarVisita(Data, Horario),
    (Horario = '0' -> writeln('Nao existe mais vaga neste dia!');
    (atomic_list_concat(['\nVoce foi agendado para ', Data, ' as ', Horario, '!'], String),
    write(String))).

%---------------------------- Métodos Extra ---------------------------

toString(List, String) :-
    nth0(0, List, Data),
    nth0(1, List, Horario1),
    nth0(2, List, Horario2),
    nth0(3, List, Horario3),
    nth0(4, List, Horario4),
    nth0(5, List, Horario5),
    Lista = ['---------------------------------- ', Data, ' ----------------------------------\n', '8h as 10h: ', Horario1, ' | 10h as 12h: ', Horario2, ' | 12h as 14h: ', Horario3, ' | 14h as 16h: ', Horario4, ' | 16h as 18h: ', Horario5],
    atomic_list_concat(Lista, String).


seNaoTemAgendas:-
    lerCsvRowList('../db/Agendas.csv', Agendas),
    proper_length(Agendas, Tamanho),
    Tamanho =:= 0.